framework/src/eigensolver/monolith.c

```diff
@@ -0,0 +1,559 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#if LIBMESH_HAVE_SLEPC
+
+#include <slepc/private/epsimpl.h> /*I "slepceps.h" I*/
+#include <slepcblaslapack.h>
+#include <petscsnes.h>
+
+typedef struct {
```

Comment:
do you have any pdf document i can read to get a better idea on the code? Looks good. This is something I want, move iteration inside C codes and let MOOSE only provide callbacks. BTW, where is the free power iterations for the nonlinear eigen solve?